lib/cusparse/generic.jl

```diff
@@ -196,93 +166,64 @@ function mv!(transa::SparseChar, alpha::Number, A::CuSparseMatrixCSC{TA}, X::Den
 
     function bufferSize()
         out = Ref{Csize_t}()
-        cusparseSpMV_bufferSize(handle(), ctransa, Ref{compute_type}(alpha), descA, descX, Ref{compute_type}(beta),
+        cusparseSpMV_bufferSize(handle(), transa2, Ref{compute_type}(alpha), descA, descX, Ref{compute_type}(beta),
                                 descY, compute_type, algo, out)
         return out[]
     end
     with_workspace(bufferSize) do buffer
-        cusparseSpMV(handle(), ctransa, Ref{compute_type}(alpha), descA, descX, Ref{compute_type}(beta),
+        cusparseSpMV(handle(), transa2, Ref{compute_type}(alpha), descA, descX, Ref{compute_type}(beta),
                      descY, compute_type, algo, buffer)
     end
-
     return Y
 end
 
-function mm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::CuSparseMatrixCSR{T},
+function mm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::Union{CuSparseMatrixCSC{T},CuSparseMatrixCSR{T},CuSparseMatrixCOO{T}},
              B::DenseCuMatrix{T}, beta::Number, C::DenseCuMatrix{T}, index::SparseChar, algo::cusparseSpMMAlg_t=CUSPARSE_MM_ALG_DEFAULT) where {T}
-    m,k = size(A)
-    n = size(C)[2]
-
-    if transa == 'N' && transb == 'N'
-        chkmmdims(B,C,k,n,m,n)
-    elseif transa == 'N' && transb != 'N'
-        chkmmdims(B,C,n,k,m,n)
-    elseif transa != 'N' && transb == 'N'
-        chkmmdims(B,C,m,n,k,n)
-    elseif transa != 'N' && transb != 'N'
-        chkmmdims(B,C,n,m,k,n)
-    end
-
-    descA = CuSparseMatrixDescriptor(A, index)
-    descB = CuDenseMatrixDescriptor(B)
-    descC = CuDenseMatrixDescriptor(C)
-
-    function bufferSize()
-        out = Ref{Csize_t}()
-        cusparseSpMM_bufferSize(
-            handle(), transa, transb, Ref{T}(alpha), descA, descB, Ref{T}(beta),
-            descC, T, algo, out)
-        return out[]
-    end
-    with_workspace(bufferSize) do buffer
-        cusparseSpMM(
-            handle(), transa, transb, Ref{T}(alpha), descA, descB, Ref{T}(beta),
-            descC, T, algo, buffer)
-    end
 
-    return C
-end
-
-function mm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::CuSparseMatrixCSC{T},
-             B::DenseCuMatrix{T}, beta::Number, C::DenseCuMatrix{T}, index::SparseChar, algo::cusparseSpMMAlg_t=CUSPARSE_MM_ALG_DEFAULT) where {T}
-    ctransa = 'N'
-    if transa == 'N'
-        ctransa = 'T'
-    elseif transa == 'C' && T <: Complex
+    if isa(A, CuSparseMatrixCSC) && transa == 'C' && T <: Complex
         throw(ArgumentError("Matrix-matrix multiplication with the adjoint of a CSC matrix" *
```

Comment:
The error message should now mention Complex, right? Also, doesn't the code below convert to CSR by setting `convert=true`?

Reply:
`transa='C'` is only supported when the matrix is complex.
Hum, we could set transa2='T' if `transa='C'` and `T <: Real` in sv!, sm!, mv!, mm! routines. :thinking: 
I will add this option and update the error message.

The code below convert to CSR by setting `convert=true` only if `A` is real.
We have `conj(A)` if `A` is complex because the trick with the swap of pointers is equivalent to transpose.

Reply:
I updated the error message and also added the support of `transa='C' for real matrices.